### PR TITLE
#19489 #19458 Unique fields and tag fields are mapped as keywords

### LINKED ISSUE #19458: [Content] : Tag field from user returns different data compared to system Tag field
Reported via Support: [102600](https://dotcms.zendesk.com/agent/tickets/102600)
Reproducible **as of dotCMS 5.2.7** up to latest master.

When using a Lucene query to return data, the value of the `+$contentTypeVarName.tagFieldVarName:{tag-name}` causes the query to return different results compared to using the system Tags field:`+tags:{tag-name}`

**To Reproduce**
1. Create a test Content Type called `Test CT` with two fields:
1.1 Text field called Title, required, searchable.
1.2 Tag field called My Tags, required, searchable.
2. Create these two tags:
2.1 Name: `|africa`
2.2 Name: `|africa|algeria`
3. Create two Contentlets of type `Test CT`, one pointing to the `|africa` tag, and one pointing to `|africa|algeria`
4. In the **Content Search** portlet, filter the contents by **Tag** with this value: `|africa`. However, it returns two both contents instead of just one:
query-content-search

The **Show Query** modal indicates the Lucene query used to retrieve the data:
```
+contentType:TestCt +(conhost:48190c8c-42c4-46af-8d1a-0cd5db894797 conhost:SYSTEM_HOST) +TestCt.myTags:"\|africa" +languageId:1 +deleted:false +working:true
```
5. If you copy the query above, go to the **Query Tool** portlet and replace `+TestCt.myTags:"\|africa"` with `+tags:"\|africa"`, you will get the expected behavior now:
query-tool-data

**Expected behavior**
Using either tag reference must return the same data, as no wildcards are being passed down to the value in order to retrieve more than one result.

**Additional context**
The currently valid workaround is to use the `+tags:{tag-name}` parameter to query for one specific Tag name, and not to get contents whose Tags start with `{tag-name}`

### LINKED ISSUE #19489: If a field is marked unique, we should es map it as keyword
When we checkin a contentlet that has a unique field that is stored as text we need to map it in elasticsearch as a keyword.
This is because when we validate a unique record, we do an elasticsearch query against the value of the field.  When the key contains a space or a dash, this unique lookup will fail.

To reproduce
1. Create a content type with a single unique field
2. Add one contentlet with the value of this field 'aaa-bbb-ccc'
3. Add one contentlet with the value of this field 'aaa-bbb-ddd'

### LINKED ISSUE #19536: CSV Importer should use _dotraw for text key and unique fields
If a user selects a key field while importing, the importer tool should use the _dotraw when performing the query to check if the field exists, etc.

Likewise for content types with unique fields, we should use _dotraw to validate a content's uniqueness 

This is a performance issue - without using the _dotraw, the importer can be slow and the uniqueness not return accurate results.

### PR DESCRIPTION
**Pull request by nollymar**
With these changes, unique text fields are mapped in ES as keywords. 
On the other hand, lucene queries used to validate fields uniqueness use _dotraw fields except for unique numbers (integer and float fields), as numeric datatypes behave like keywords in searches.

### CODE REVIEW

**Review by dsilvam — APPROVED** [review]

**Review by wezell — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java [hunk 1]
```diff
@@ -361,20 +363,31 @@ private void addMappingForFieldIfNeeded(
...
+                    || field instanceof WysiwygField || field instanceof RadioField
+                    || field instanceof SelectField || field instanceof MultiSelectField
+                    || field instanceof TagField) {
+
+                if (dataTypesMap.containsKey(field.dataType())) {
+                    mappingForField = String
+                            .format("{\n\"type\":\"%s\"\n}",
+                                    dataTypesMap.get(field.dataType()));
+                } else {
+                    if (field.unique() || field instanceof TagField) {
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:382]: @wezell it's already implemented here: https://github.com/dotCMS/core/blob/master/dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESMappingAPIImpl.java#L703-L708 . But I'll confirm with Kibana 👍

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 2]
```diff
@@ -6311,9 +6313,18 @@ else if(field.getFieldType().equals(Field.FieldType.DATE_TIME.toString())){
...
+                    buffy.append(" +" + contentlet.getContentType().variable()).append(StringPool.PERIOD)
+                            .append(field.getVelocityVarName());
+
+                    if (isDataTypeNumber){
+                        buffy.append(StringPool.COLON).append("\"").append(getFieldValue(contentlet,
+                                new LegacyFieldTransformer(field).from())).append("\"");
+                    }else{
+                        buffy.append("_dotraw").append(StringPool.COLON).append("\"")
+                                .append(getFieldValue(contentlet,
+                                        new LegacyFieldTransformer(field).from()))
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:6325]: I'll do some tests

**Review by fabrizzio-dotCMS — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:382]: I have confirmed. No additional change is needed

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: @wezell I have fixed it. I modified the logic to use the `_sha256` field to make the validation more accurate and consider special characters. 

Besides, `_sha256` are now mapped as `keyword`

**Pull request merged**